merkledag/utils/utils.go

```diff
@@ -0,0 +1,113 @@
+package dagutils
+
+import (
+	"errors"
+	"time"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+func AddLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname string, childk key.Key) (*dag.Node, error) {
+	if childname == "" {
+		return nil, errors.New("cannot create link with no name!")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+	childnd, err := ds.Get(ctx, childk)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure no link with that name already exists
+	_ = root.RemoveNodeLink(childname) // ignore error, only option is ErrNotFound
+
+	err = root.AddNodeLinkClean(childname, childnd)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = ds.Add(root)
+	if err != nil {
+		return nil, err
+	}
+	return root, nil
+}
+
+func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key, create bool) (*dag.Node, error) {
```

Comment:
why not make path a ... "/" separated path? just because of the recursion?  it's probably a better UX to use paths. can use a helper function to do the recursion:

```go
func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path string, toinsert key.Key, create bool) (*dag.Node, error) {
  p := strings.Split(path, "/")
  return insertNodeAtPath(ctx, ds, root, p, toinsert, create)
}
```

also, these functions with a bunch of params could use a helper type:

```go
type Editor struct {
  DS     dag.DAGService
  Root   dag.Node
  Create bool
}

func (de *Editor) InsertNodeAtPath(ctx context.Context, p []string, k key.Key) (*dag.Node, error) { ... }
```

The caller becomes clearer

```go
func main() {
  ed := dagutil.Editor{
    DS: dagService,
    Root: nd,
    Create: true,
  }

  if _, err := ed.InsertNodeAtPath(ctx, "a/b/c", c); err != nil { ... }
  if _, err := ed.InsertNodeAtPath(ctx, "a/b/d", d); err != nil { ... }
  ...
}
```